res/controllers/midi-components-0.1.js

```diff
@@ -0,0 +1,1136 @@
+        outSetParameter(value) {
+            engine.setParameter(this.group, this.outKey, value);
+        }
+        outGetValue() {
+            return engine.getValue(this.group, this.outKey);
+        }
+        outSetValue(value) {
+            engine.setValue(this.group, this.outKey, value);
+        }
+        outToggle() {
+            this.outSetValue(!this.outGetValue());
+        }
+
+        connect() {
+            /**
+            Override this method with a custom one to connect multiple Mixxx COs for a single Component.
+            Add the connection objects to the this.connections array so they all get disconnected just
+            by calling this.disconnect(). This can be helpful for multicolor LEDs that show a
+            different color depending on the state of different Mixxx COs. Refer to
+            SamplerButton.connect() and SamplerButton.output() for an example.
+            **/
+            if (undefined !== this.group &&
+                undefined !== this.outKey &&
+                undefined !== this.output &&
+                typeof this.output === "function") {
+                this.connections[0] = engine.makeConnection(this.group, this.outKey, this.output.bind(this));
+            }
+        }
+        disconnect() {
+            if (this.connections[0] !== undefined) {
+                this.connections.forEach(function(conn) {
+                    conn.disconnect();
+                });
+            }
+        }
+        trigger() {
+            if (this.connections[0] !== undefined) {
+                this.connections.forEach(function(conn) {
+                    conn.trigger();
+                });
+            }
+        }
+        send(value) {
+            if (this.midi === undefined || this.midi[0] === undefined || this.midi[1] === undefined) {
+                return;
+            }
+            midi.sendShortMsg(this.midi[0], this.midi[1], value);
+            if (this.sendShifted) {
+                if (this.shiftChannel) {
+                    midi.sendShortMsg(this.midi[0] + this.shiftOffset, this.midi[1], value);
+                } else if (this.shiftControl) {
+                    midi.sendShortMsg(this.midi[0], this.midi[1] + this.shiftOffset, value);
+                }
+            }
+        }
+    };
+    Component.prototype.max = 127;
+    Component.prototype.shiftOffset = 0;
+    Component.prototype.sendShifted = false;
+    Component.prototype.shiftChannel = false;
+    Component.prototype.shiftControl = false;
+
+    var Button = class extends Component {
+        constructor(options) {
+            if (options.type === undefined) {
+                options.type = Button.prototype.types.push;
+            }
+            super(options);
+        }
+        isPress(channel, control, value, _status) {
+            return value > 0;
+        }
+        input(channel, control, value, status, _group) {
+            if (this.type === undefined || this.type === Button.prototype.types.push) {
```

Comment:
type can never be undefined right? Please use a switch case and throw an error in the default case.

Reply:
This is JavaScript. Everything and anything can be undefined.

Reply:
It is defined in the constructor though